File: CS/T179722/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraEditors;
using DevExpress.Skins;
using DevExpress.LookAndFeel;
using DevExpress.UserSkins;
using DevExpress.XtraScheduler;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;
using System.Collections;

namespace T179722 {
    public partial class Form1 : XtraForm {

        public Form1() {
            InitializeComponent();
            this.schedulerControl.Start = DateTime.Now;
        }

        GridHitInfo DownHitInfo { get; set; }

        void OnForm1Load(object sender, EventArgs e) {
            this.schedulerControl.GroupType = SchedulerGroupType.Resource;
            this.schedulerStorage.Appointments.ResourceSharing = true;

            DataHelper.FillResources(this.schedulerStorage, 5);
            this.grdTasks.DataSource = DataHelper.GenerateScheduleTasks();
        }

        void OnGridViewTasksMouseDown(object sender, MouseEventArgs e) {
            GridView view = sender as GridView;
            this.DownHitInfo = null;

            GridHitInfo hitInfo = view.CalcHitInfo(new Point(e.X, e.Y));
            if (Control.ModifierKeys != Keys.None)
                return;
            if (e.Button == MouseButtons.Left && hitInfo.InRow && hitInfo.HitTest != GridHitTest.RowIndicator)
                this.DownHitInfo = hitInfo;
        }

        void OnGridViewTasksMouseMove(object sender, MouseEventArgs e) {
            GridView view = sender as GridView;
            if (e.Button == MouseButtons.Left && this.DownHitInfo != null) {
                Size dragSize = SystemInformation.DragSize;
                Rectangle dragRect = new Rectangle(new Point(this.DownHitInfo.HitPoint.X - dragSize.Width / 2,
                    this.DownHitInfo.HitPoint.Y - dragSize.Height / 2), dragSize);

                if (!dragRect.Contains(new Point(e.X, e.Y))) {
                    view.GridControl.DoDragDrop(GetDragData(view), DragDropEffects.All);
                    this.DownHitInfo = null;
                }
            }
        }

        void OnSchedulerControlAppointmentDrop(object sender, AppointmentDragEventArgs e) {
            string createEventMsg = "Creating an event at {0} on {1}.";
            string moveEventMsg = "Moving the event from {0} on {1} to {2} on {3}.";

            DateTime srcStart = e.SourceAppointment.Start;
            DateTime newStart = e.EditedAppointment.Start;

            string msg = (srcStart == DateTime.MinValue) ? String.Format(createEventMsg, newStart.ToShortTimeString(), newStart.ToShortDateString()) :
                String.Format(moveEventMsg, srcStart.ToShortTimeString(), srcStart.ToShortDateString(), newStart.ToShortTimeString(), newStart.ToShortDateString());

            if (XtraMessageBox.Show(msg + "\r\nProceed?", Application.ProductName, MessageBoxButtons.YesNo, MessageBoxIcon.Question) == DialogResult.No) {
                e.Allow = false;
            }
        }

        void OnSchedulerControlAppointmentDrag(object sender, AppointmentDragEventArgs e) {
            Appointment apt = e.EditedAppointment;
            if (apt.Start == DateTime.MinValue)
                apt.Start = e.HitInterval.Start;
        }

        void OnSchedulerControlAdditionalAppointmentsDrag(object sender, AdditionalAppointmentsDragEventArgs e) {
            List<AppointmentExchangeData> exchangeList = new List<AppointmentExchangeData>();
            foreach (AppointmentDragInfo aptInfo in e.PrimaryAppointmentInfos) {
                var apt = aptInfo.EditedAppointment;
                exchangeList.Add(new AppointmentExchangeData() {
                    Subject = apt.Subject,
                    Description = apt.Description,
                    Start = DateTime.MinValue,
                    Duration = apt.Duration,
                    LabelKey = (int)apt.LabelKey,
                    StatusKey = (int)apt.StatusKey
                });
            }
            e.Data.SetData(DataFormats.Serializable, exchangeList);
        }

        void OnSchedulerControlPrepareDragData(object sender, PrepareDragDataEventArgs e) {
            object data = e.DataObject.GetData(DataFormats.Serializable);
            AppointmentBaseCollection appointments = new AppointmentBaseCollection();
            foreach (AppointmentExchangeData item in (IList)data) {
                var apt = this.schedulerStorage.CreateAppointment(AppointmentType.Normal);
                apt.Subject = item.Subject;
                apt.Description = item.Description;
                apt.Start = item.Start;
                apt.Duration = item.Duration;
                apt.LabelKey = item.LabelKey;
                apt.StatusKey = item.StatusKey;
                appointments.Add(apt);
            }
            SchedulerDragData schedulerDragData = new SchedulerDragData(appointments);
            e.DragData = schedulerDragData;
        }

        IDataObject GetDragData(GridView view) {
            int[] selection = view.GetSelectedRows();
            if (selection == null)
                return null;

            List<AppointmentExchangeData> exchangeList = new List<AppointmentExchangeData>();
            int count = selection.Length;
            for (int i = 0; i < count; i++) {
                int rowIndex = selection[i];
                exchangeList.Add(new AppointmentExchangeData() {
                    Subject = (string)view.GetRowCellValue(rowIndex, "Subject"),
                    LabelKey = (int)view.GetRowCellValue(rowIndex, "Severity"),
                    StatusKey = (int)view.GetRowCellValue(rowIndex, "Priority"),
                    Start = DateTime.MinValue,
                    Duration = TimeSpan.FromHours((int)view.GetRowCellValue(rowIndex, "Duration")),
                    Description = (string)view.GetRowCellValue(rowIndex, "Description"),
                });
            }

            return new DataObject(DataFormats.Serializable, exchangeList);
        }
    }
}

File: VB/T179722/Form1.vb
Imports System
Imports System.ComponentModel
Imports System.Drawing
Imports System.Windows.Forms
Imports DevExpress.XtraEditors
Imports DevExpress.Skins
Imports DevExpress.XtraScheduler
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo

Namespace T179722

    Public Partial Class Form1
        Inherits XtraForm

        Public Sub New()
            InitializeComponent()
            schedulerControl.Start = Date.Now
        End Sub

        Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs)
            schedulerControl.GroupType = SchedulerGroupType.Resource
            schedulerStorage.Appointments.ResourceSharing = True
            DataHelper.FillResources(schedulerStorage, 5)
            grdTasks.DataSource = DataHelper.GenerateScheduleTasks()
        End Sub

'#Region "#GetGridHitInfo"
        Private downHitInfo As GridHitInfo

        Private Sub gridViewTasks_MouseDown(ByVal sender As Object, ByVal e As MouseEventArgs)
            Dim view As GridView = TryCast(sender, GridView)
            downHitInfo = Nothing
            Dim hitInfo As GridHitInfo = view.CalcHitInfo(New Point(e.X, e.Y))
            If ModifierKeys <> Keys.None Then Return
            If e.Button = MouseButtons.Left AndAlso hitInfo.InRow AndAlso hitInfo.HitTest <> GridHitTest.RowIndicator Then downHitInfo = hitInfo
        End Sub

'#End Region  ' #GetGridHitInfo
'#Region "#GridViewMouseMove"
        Private Sub gridViewTasks_MouseMove(ByVal sender As Object, ByVal e As MouseEventArgs)
            Dim view As GridView = TryCast(sender, GridView)
            If e.Button = MouseButtons.Left AndAlso downHitInfo IsNot Nothing Then
                Dim dragSize As Size = SystemInformation.DragSize
                Dim dragRect As Rectangle = New Rectangle(New Point(downHitInfo.HitPoint.X - dragSize.Width \ 2, downHitInfo.HitPoint.Y - dragSize.Height \ 2), dragSize)
                If Not dragRect.Contains(New Point(e.X, e.Y)) Then
                    view.GridControl.DoDragDrop(GetDragData(view), DragDropEffects.All)
                    downHitInfo = Nothing
                End If
            End If
        End Sub

'#End Region  ' #GridViewMouseMove
'#Region "#GetDragData"
        Private Function GetDragData(ByVal view As GridView) As SchedulerDragData
            Dim selection As Integer() = view.GetSelectedRows()
            If selection Is Nothing Then Return Nothing
            Dim appointments As AppointmentBaseCollection = New AppointmentBaseCollection()
            Dim count As Integer = selection.Length
            For i As Integer = 0 To count - 1
                Dim rowIndex As Integer = selection(i)
                Dim apt As Appointment = schedulerStorage.CreateAppointment(AppointmentType.Normal)
                apt.Subject = CStr(view.GetRowCellValue(rowIndex, "Subject"))
                apt.LabelId = CInt(view.GetRowCellValue(rowIndex, "Severity"))
                apt.StatusId = CInt(view.GetRowCellValue(rowIndex, "Priority"))
                apt.Start = Date.Now
                apt.Duration = TimeSpan.FromHours(CInt(view.GetRowCellValue(rowIndex, "Duration")))
                apt.Description = CStr(view.GetRowCellValue(rowIndex, "Description"))
                appointments.Add(apt)
            Next

            Return New SchedulerDragData(appointments, 0)
        End Function

'#End Region  ' #GetDragData
'#Region "#SchedulerControlAppointmentDrop"
        Private Sub schedulerControl_AppointmentDrop(ByVal sender As Object, ByVal e As AppointmentDragEventArgs)
            Dim createEventMsg As String = "Creating an event at {0} on {1}."
            Dim moveEventMsg As String = "Moving the event from {0} on {1} to {2} on {3}."
            Dim srcStart As Date = e.SourceAppointment.Start
            Dim newStart As Date = e.EditedAppointment.Start
            Dim msg As String = If(srcStart = Date.MinValue, String.Format(createEventMsg, newStart.ToShortTimeString(), newStart.ToShortDateString()), String.Format(moveEventMsg, srcStart.ToShortTimeString(), srcStart.ToShortDateString(), newStart.ToShortTimeString(), newStart.ToShortDateString()))
            If XtraMessageBox.Show(msg & Microsoft.VisualBasic.Constants.vbCrLf & "Proceed?", Application.ProductName, MessageBoxButtons.YesNo, MessageBoxIcon.Question) = DialogResult.No Then
                e.Allow = False
            End If
        End Sub
'#End Region  ' #SchedulerControlAppointmentDrop
    End Class
End Namespace
